pulsar-common/src/main/java/org/apache/pulsar/common/api/raw/MessageParser.java

```diff
@@ -94,7 +94,7 @@ public static void parseMessage(TopicName topicName, long ledgerId, long entryId
                 throw new IOException("Cannot parse encrypted message " + msgMetadata + " on topic " + topicName);
             }
 
-            uncompressedPayload = uncompressPayloadIfNeeded(topicName, msgMetadata, headersAndPayload, ledgerId,
+            uncompressedPayload = uncompressPayloadIfNeeded(topicName, msgMetadata, headersAndPayload.retain(), ledgerId,
```

Comment:
Probably better to let `ReferenceCountedMessageMetadata` do both `retain` and `release`, else caller forget to call `retain` might cause trouble when `ReferenceCountedMessageMetadata` try to do `release`.

Reply:
Good point, updated.